### PR TITLE
Adding Extension metadata

### PR DESCRIPTION
**Pull request by carlopi**
### Why?
* Observability of what extension you have at the moment on your system
* Allow checks on consistency of both duckdb's version and platform before actually loading a binary

Both together allows to avoid a class of bugs where if for any reason duckdb was instructed to load a given extension (say via the explicit LOAD 'path/to/file'), and there was a platform mismatch this lead to weird and complex to debug crashes.

Also, observability of current available extensions allow more precise informations while tracking for bugs related to extensions.

### How
Extensions binary now have a 512 bytes payload added at the end. Of those 256 are reserved for metadata and 256 for the signature.

Of the 256 metadata bytes only 4 slots of 32 bytes are actually used, with 128 bytes reserved for future usage.
The current 4 slots are:
* A slot that defaults to "4". This serve as the metadata version, and currently doubles as the number of available slots, but meaning can and will change
* A slot for the duckdb_platform the extension has been build for
* A slot for the duckdb_version the extension has been build for
* A slot for the extension_version of the extension. Currently can be either
  - a string containing a git reference (either tag or hash) for out of tree extensions
  - the string "in-tree" for in-tree extensions
  - a custom string if build_loadable_extension_directory is invoked directly, see test/extension/CMakeLists.txt where for loadable_extension_demo is set to "default-version"

Given duckdb-wasm extensions need to have the payload be part of a [custom section](https://webassembly.github.io/spec/core/appendix/custom.html), a set of magic bytes are added before the 512 bytes payload, and those allows the script to be equivalent between wasm & native cases, and has the property that given in input a valid WebAssembly module a valid WebAssembly module will be returned in output. Custom section specification encodes basically the length of the segment that will follow.

NOTE: Signing script has also been changed, now any extension build by duckdb will have an empty signature, so when extension are signed first file need to be truncated and only then signature can be added.

### Demo
If you don't want to wait for CI to try this out, steps are like:

1. Add those lines at the bottom of `extension/extension_config.cmake`
```
duckdb_extension_load(json)
duckdb_extension_load(autocomplete)
duckdb_extension_load(icu DONT_LINK)
duckdb_extension_load(httpfs DONT_LINK)
duckdb_extension_load(tpch DONT_LINK)
duckdb_extension_load(tpcds DONT_LINK)
duckdb_extension_load(sqlite_scanner
        DONT_LINK LOAD_TESTS
        GIT_URL https://github.com/duckdb/sqlite_scanner
        GIT_TAG 70a4c411434290b4af704940987a324d84af0024
        APPLY_PATCHES
        )
    duckdb_extension_load(postgres_scanner
            DONT_LINK
            GIT_URL https://github.com/duckdb/postgres_scanner
            GIT_TAG 375710fd22a35107b2c28e744f787e1a93a99998
            )
```
Then run:
```sql
make clean
# build DuckDB in a setting similar to CI
ENABLE_EXTENSION_AUTOLOADING=1 ENABLE_EXTENSION_AUTOINSTALL=1  GEN=ninja make
# launch DuckDB with unsigned + instructing it to fetch extension from the default local extension repository
./build/release/duckdb -unsigned -cmd "SET custom_extension_repository='build/release/repository';"
CREATE TABLE services AS FROM 'https://blobs.duckdb.org/stations.csv.gz'; --- autoinstall+load httpfs
ATTACH 'hi.db' (TYPE SQLITE);  --- autoinstall+load sqlite
INSTALL postgres; LOAD postgres;   --- explicit install + loading
SELECT * EXCLUDE(description, install_path, loaded, aliases) FROM duckdb_extensions();
┌──────────────────┬───────────┬──────────────────────────────────┐
│  extension_name  │ installed │        extension_version         │
│     varchar      │  boolean  │             varchar              │
├──────────────────┼───────────┼──────────────────────────────────┤
│ arrow            │ false     │                                  │
│ autocomplete     │ true      │                                  │
│ aws              │ false     │                                  │
│ azure            │ false     │                                  │
│ excel            │ false     │                                  │
│ fts              │ false     │                                  │
│ httpfs           │ true      │ in-tree                          │
│ iceberg          │ false     │                                  │
│ icu              │ false     │                                  │
│ inet             │ false     │                                  │
│ jemalloc         │ false     │                                  │
│ json             │ true      │                                  │
│ motherduck       │ false     │                                  │
│ mysql_scanner    │ false     │                                  │
│ parquet          │ true      │                                  │
│ postgres_scanner │ true      │ ce22bc41da1b876a8d541e013251712c │
│ shell            │           │                                  │
│ spatial          │ false     │                                  │
│ sqlite_scanner   │ true      │ 9b558ed2e933817bff96726fec0868e7 │
│ substrait        │ false     │                                  │
│ tpcds            │ false     │                                  │
│ tpch             │ false     │                                  │
│ visualizer       │ false     │                                  │
├──────────────────┴───────────┴──────────────────────────────────┤
│ 23 rows                                               3 columns │
└─────────────────────────────────────────────────────────────────┘
```
Up to here the same behaviour would be true also with main, but the extension_version column would be missing.

Current duckdb:
```
D LOAD 'build/release/duckdb';
Error: IO Error: File "build/release/duckdb" did not contain function "duckdb_version": dlsym(0x97b3b5d0, duckdb_version): symbol not found
```
After this PR:
```
D LOAD 'build/release/duckdb';
Error: Invalid Input Error: Extension "build/release/duckdb" (version X?F?`?.?D?
?uR?,??B?^V>H?2L) does not match DuckDB loading it (version 4d67790aef, platform osx_arm64)Ln??	??, platfrom ?^?8?$?>?;
```
(just realized I need to invert logic of errors, and fail explaining it do not seem to be a duckdb extension at all, and not printing random bytes from a file)
After this PR + an extra commit:
```
D LOAD 'build/release/duckdb';
Error: Invalid Input Error: Extension "build/release/duckdb" do not have metadata compatible with DuckDB version 4d67790aef```
```

### Changes required for users / extension developers
* extension_version function can be removed/deprecated from extensions as not used anymore
* extensions will need to use the new script to be updated, after this PR I will need to change duckdb's extensions CI to rely on new metadata generation & signing script
* `FROM duckdb_extensions();` now returns an additional column, the extension_version. If somehow relies on exact schema, that will need to be adapted.

### Up next
PR https://github.com/duckdb/duckdb/pull/10437 is connected to this one, and would allow the INSTALL of a specific extension version.
Currently extension_version is correctly propagated only for loaded (non-built-in) extensions, I will to handle also installed extension via the same logic to be used in PR https://github.com/duckdb/duckdb/pull/10437.

### CODE REVIEW

File: scripts/extension-upload-single.sh [hunk 1]
```diff
@@ -33,21 +33,7 @@ script_dir="$(dirname "$(readlink -f "$0")")"
...
-  # 10 in hex = 16 in decimal, lenght of name, 1 byte
-  echo -n -e '\x10' >> $ext.append
-  echo -n -e 'duckdb_signature' >> $ext.append
-  # the name of the WebAssembly custom section, 16 bytes
-  # 100 in hex, 256 in decimal
-  # [1(continuation) + 0000000(payload) = ff, 0(continuation) + 10(payload)],
-  # for a grand total of 2 bytes
-  echo -n -e '\x80\x02' >> $ext.append
-fi
+(( command -v truncate && truncate -s -256 $ext.append )) || ((command -v gtruncate && gtruncate -s -256 $ext.append )) || exit 1
```
Comment (samansmink) [inline scripts/extension-upload-single.sh:36]: I'm getting a syntax error here:
```
/Users/sam/Development/duckdb/scripts/extension-upload-single.sh: line 36: ((: command -v truncate && truncate -s -256 build/release/extension/autocomplete/autocomplete.duckdb_extension.append : syntax error in expression (error token is "truncate && truncate -s -256 build/release/extension/autocomplete/autocomplete.duckdb_extension.append ")
```

File: src/main/extension/extension_load.cpp [hunk 2]
```diff
@@ -3,6 +3,7 @@
 #include "duckdb/main/extension_helper.hpp"
 #include "duckdb/main/error_manager.hpp"
 #include "mbedtls_wrapper.hpp"
+//#include <iostream>
```
Comment (samansmink) [inline src/main/extension/extension_load.cpp:6]: nit: can be removed

File: src/main/extension/extension_load.cpp [hunk 3]
```diff
@@ -119,15 +120,50 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileSystem &fs, const str
 		error = StringUtil::Format("Extension \"%s\" not found.\n%s", filename, message);
 		return false;
 	}
-	if (!config.options.allow_unsigned_extensions) {
-		auto handle = fs.OpenFile(filename, FileFlags::FILE_FLAGS_READ);
 
-		// signature is the last 256 bytes of the file
+	string metadata_segment;
```
Comment (samansmink) [inline src/main/extension/extension_load.cpp:124]: I think the `ExtensionHelper::TryInitialLoad` has reached a point here where it is a bit too long to be readable, could we split this one up in some smaller functions?

**Review by samansmink — COMMENTED** [review]
Hey @carlopi! I took a look here. I have some minor issues, and 1 bigger one where I think the upload script will fail due to a syntax error

File: scripts/extension-upload-single.sh (hunk 1 above)
Comment (carlopi) [reply scripts/extension-upload-single.sh:36]: Thanks for double checking this, resolved.

**Comment by carlopi** [comment]
Closing this in favor of https://github.com/duckdb/duckdb/pull/11515, a refactor with reworked error messages, cleanup and tests.

**Pull request closed without merging**